### ISSUE TITLE
Data Mapping to records. Which layer.

### ISSUE THREAD

**Issue by dmill-bz** [spider/spider#11]
Here's a recap of previous conversations:

---

The mapping of DB results to objects in the driver just seems out of place. It shouldn't be the driver's place to know what the query is returning. 
In the following case:

``` php
$query = UserRecord::select();
$userRecords = $query->all();
```

It's the Query that gets the `all()`, `one()`, `tree()`, etc. calls. It should be the query's job to populate the records (as it also knows which record spawned it). I think the `SpiderRecord` in the driver is too premature, or at least not properly implemented. 

Maybe the `Driver` should implement multiple formatting methods for different result sets. path, tree, vertices, properties, scalar and the query should call those appropriately.

For instance, `$query->all()` would in reality do:

``` php
$result = $query->execute();
$result = Driver::FormatAll($result);
// Do mapping of results to objects
$return $objectArray;
```

(Doesn't have to be exactly like this but it's for the idea)

---

**Mapping Results**

Mapping results to `SpiderRecord`s in the `Driver` has always felt awkward to me. I've tried it in other places like the `Connection` and on its own.

The problem is twofold. First, people should be able to use these component parts independently. Different layers of abstraction will build on top of each other. That is, use the `Driver` without using the `CommandBuilder`. Once Models come around, they should be able to use the `CommandBuilder` without using the `Driver` directly.

With that in mind, the `Driver` MUST return something consistent. In the orient driver, I am using a third-party language binding and I expect most drivers to do the same. Those bindings will each return their own flavor of records. If it is the `CommandBuilder` or `Connection` or something else that is responsible for mapping, I don't know how to consistently map an unknown record type to a consistent SpiderRecord.

Maybe each driver could/should contain a `RecordTransformer` of some kind? If this one needs more discussion, lets open a new issue and copy some of this.

**Comment by dmill-bz** [comment]
I guess a few questions need answers.

**What are the most common response types?**
- Set of elements
- Single element (should generally be the same as above just with one entry)
- Set of properties (single element)
- Set of set of properties (multiple elements)
- Tree
- Path
- Scalar
- Custom

**Which of these should we handle (ignoring the others)?**
- Set of elements
- Single element (should generally be the same as above just with one entry)
- Tree
- Path
- Scalar

**We need to have the Driver return consistent data across DBs but what type is that?**

Should it be a PHP native formatted in a specific way? I don't think it should be a `Record` because the Driver could be used on it's own without records.
Also, the Command spawned by a `Record` could delegate the mapping to the `Record` that spawned it (if it exists.. like say `UserRecord`). This way we properly abstract all layers.

Scenarios would be:

```
String Query -> Driver -> PHP native response
Command -> String Query -> Driver -> PHP native response -> Command -> returns PHP native
Record -> Command -> String Query -> Driver -> PHP native response -> Command -> has Record and uses it to map response to Records
```

Just some thinking out loud.

**Comment by electricjones** [comment]
So are you equating `Record`s with the Models found in most ORMs? I was considering a Model (yet to be implemented) as an _extension_ of a Record. The Record itself is just a plain-php-object with the properties (and maybe a few helper methods for accessors, mutators). It's mostly for typehinting.

If the `Driver` will be returning a plain php object, that actually means that many drivers who use third party language bindings will have to strip the data out of whatever the binding returns. The orientdb binding actually returns and `PhpOrientRecord` object that is more convoluted than it needs to be. I guess that's what the mapper does already, anyway.

So, I propose the driver always returns a `scalar` (for a single property), a plain php object (for a single database record like a vertex), or an `array` for a set of single records.

The `CommandBuilder` should be responsible for mapping those result sets to trees, paths, and the like.

It may be nice for the plain-php-objects returned by the Drivers to be of a certain class for type-hinting.

**Comment by dmill-bz** [comment]
Yeah that's how I've been considering `Record` (or to be more specific `Vertex` and `Edge`). Even if the models extend these they're still the O and should be excluded from the Driver layer.

If the drivers are to work on their own they should return a consistent response that isn't either of those. Or at least that's how I feel. I mentioned php natives but that's not necessarily the only way to go (and probably not the best).
We could look back at a `SpiderResponse` class again. The class would hold the original response as well as a reference to the driver and use that driver to format that original response to consistent formats on demand.

`SpiderResponse::getSet()`
`SpiderResponse::getTree()`
`SpiderResponse::getPath()` 
`SpiderResponse::getScalar()` 
`SpiderResponse::getRaw()` 
etc.

These would all return php natives though (`array`, `Int`, `String`, etc..). The mapping is the responsibility of the above layers. (within the context of Spider it will be taken care of by `Record` via `CommandBuilder` but it could be anything for other systems)

This would make it easy for existing frameworks to use the drivers and language builders only and then hook those up to their own Record system.

**Comment by dmill-bz** [comment]
ok I think we're on the same wavelength here. I've just been puzzled by naming the "collection" object `Record`. Which I would attribute more to a Model.
Then in the above example `SpiderResponse` or simply `Response` would extend this `Collection` class.

**Comment by dmill-bz** [comment]
I have a bit of free time coming up I wouldn't mind helping out a bit with some commandProcessors. I think this topic needs to be clear before doing that though. 

Is it safe to assume that : 
- `Command` should not be aware of the context (not aware of either the driver or the connection) It's just a language building tool. `Command::getScript('language')`  provided the proper `CommandProcessor` exists for that language.
- `Query` is aware of context and uses (potentially wraps) `Command`
- `Query` can have a reference to a `Model` as part of it's "context". If it does it will delegate Object mapping to that `Model` and return the result instead of returning a generic `Collection` returned by the `Driver`
- `Driver`s are responsible for mapping results to a consistent format. They provide several mapping methods that the Query can use depending on the user's expected result set (`Query::all()`, `Query::one()`, `Query::tree()` etc..)
- `Model` is aware of the context, connection/driver/etc.

**Comment by electricjones** [comment]
All that sounds great. Here are my long form thoughts.

I also like where this is heading. So, let's try to nail some of this to the wall.

**First,** a `Driver` is instantiated with whatever properties it needs (as defined by the Driver Class).

``` php
class Driver extends AbstractDriver implements DriverInterface {
    public $port = 1234;
}

$driver = new Spider\Drivers\Something\Driver(['port' => 2424]);
```

**Second,** that Driver is passed an instance of a`Command`, which is just a type-hint friendly wrapper for a text query (in whatever language the driver wants)

``` php
$command = new Command("SELECT FROM Users");
$results = $driver->executeReadCommand($command);
```

**Third,** the driver executes the command and MUST return one of the following: (query builder used for illustration. Not required)
- A single, __primitive__ value if only one property was requested.

`$query->select('name')->from('users')->first();` would return `"michael"`
- An __instance of `SpiderResponse`__ for a single Vertex or Edge, analogous to a row in SQL.  `SpiderResponse` is a plain-dto used for type-hinting. If we want, it can also extend some a `Collection` for helpers. See my latest comment on #12. `SpiderResponse` has no connection back to the datastore. It is simple a result. 

`$user = $query->select()->from('users')->where('id', 1)->first();` would return a single `SpiderResponse` with `$user->name; $user->id;`
- An _array_ of `SpiderResponse`s for multiple Edges or Vertices.

`$users = $query->select()->from('users')->all()` would return an index-array with several `SpiderResponse`s

`$users = $query->select()->from('users')->where('username', 'michael')->all()` would return an index-array with a single spider response (`$users[0]`) if only one user had the username michael.

**Fourth**, any further formatting like `tree`ing or `path`ing would be handled by the `CommandBuilder`? That is the entity that is wanting specific formats.

Or, we can add a second argument to the `executeXCommand()` methods on the driver if a specific response format is requested.

`$driver->executeReadCommand($command, 'tree')` would return `SpiderTreeResponse` which extends `SpiderResponse`

The driver would be responsible for formatting these responses.

I know php doesn't support return types in interfaces, but I feel it is appropriate to expect drivers to return certain data structures and leave that implementation up to them. We can provide AbstractTransformers as a baseline. Especially if the core drivers are all in-package.

**Last**, sorry this is so long. A `SpiderResponse` could look something like this. Using magic methods to manage getting/setting. (Already taken care of if we use the `michaels/data-manager` repo spider already depends on.

``` php
$response->id; // immutable
$response->getId(); //alias
$response->getLabel(); // returns 'users', immutable

$response->name; // mutable, but not a model. Does not connect to DB
$response->whateverOtherProperties;

// Not sure the best way to handle this
$response->inEdges[0]->id; // not sure if every datastore returns the same info for queries
$response->outEdges[0]->id;

$response->meta->cluster; // meta specific to datastores
```

**Comment by dmill-bz** [comment]
This sounds good. I've got a few points I would like to clear with you still:

**Which layer decides what language a `Command`/`Query` is converted to ?**
Somehow I would expect a `Driver` to receive either a string query that is accepted "as is" or a language agnostic `Command` that it then converts to a string by choosing the language via `Command::getScript('OrientSQL')`
Driver creators can then easily define which language the driver communicates to the server with. For example:
In the event of using gremlin-server and a Neo4J backend, the driver might decide it's best to submit the query in Cypher rather than Gremlin (Both are possible but Cypher would probably be best). This is something that only the Driver would know how to do.

**SpiderResponse vs Collection**
Maybe separation of logic is in order. 
Shouldn't any response just simply return a `SpiderResponse` object. Regardless of type. Then depending on what you want from the response you can choose the appropriate format via some `getX()` method? This feels cleaner. You know to always expect a spiderResponse.

It would work like this:

``` php
$response = $driver->executeReadCommand($command);
$response instanceof SpiderResponse; // true

$response->getRaw(); // raw db data, Orient object or native php array etc...
$response->getTree(); // tree format array of Collection objects
$response->getSet(); //array of Collection objects.
// etc...
```

Formatting only happens on request. `Query` can choose the format depending on `all()`, `one()`, `tree()`, etc.

Of course if you get a tree result from the database and run `SpiderResponse::getSet()` you would get an `IncorrectFormatException` or something along those lines. But there's no avoiding this as unlike RDBMS you don't systematically get rows from your queries, so some of the burden has to fall back onto the users.

**Comment by electricjones** [comment]
To the first point, I feel the `Command` should really just hold a script and a language. The driver can check to see if that language is supported.

We already have a `CommandBag` which is a uniform, consistent way to pass commands around in an agnostic way. If a driver receives a command bag, they can simply pipe it through the CommandProcessor of its choice. So, drivers either accept a `Command` with a set script and script language, or a `Command\Bag` which it transforms into a valid `Command` object of its choice.

Completely agree with the second point.

I don't mind throwing exceptions. Never have. Better that than something unexpected.

**Comment by electricjones** [comment]
Then a `SpiderResponse` would _contain_ a specific vertex, collection of vertexes, or primitive value? Is there a way to know which of these three was returned? Or is it just up to you to know which was returned based on the query you executed? I'm good with either one.

**Comment by dmill-bz** [comment]
Ok I hear you on the `Command` stuff. That sounds good.

And the `SpiderResponse` would technically only hold the Raw data and a reference to the driver + wrapping of the driver's mapping methods.

``` php
public function getSet()
{
   return $this->_driver->FormatToSet($this->_rawData);
}
```

Something along those lines?

**Comment by dmill-bz** [comment]
Also at this stage it should be up to you to know what the result type is based on the query you submitted. This shouldn't be a problem, especially not for a first iteration.

**Comment by electricjones** [comment]
Agreed.

**Comment by dmill-bz** [comment]
Are you on this or do you want me to submit a PR for this?

**Comment by electricjones** [comment]
If you have anything started, please do. I won't be able to get back to this full swing until tomorrow, but I have all day tomorrow to focus on this repo.

Now that we have this figured out, I want to do some of that refactoring for #13 and update some of the variables and methods to reflect terminology #15.

Even if you just push a branch, that would be great. I have invited you to the repo, so you can create a branch directly from spider/spider, if that is easier.

**Comment by dmill-bz** [comment]
Yeah I've got something started. It touches quite a bit of code so I'll try and make the PR quickly

**Comment by electricjones** [comment]
Great. I'm adding `insert()`, `update()`, and `delete()` to the `CommandBuilder` today. I wouldn't imagine that would have too many conflicts. I may hold off on #11 and #12 refactoring until I see some of what you have done.

Thanks for all your help.

**Comment by dmill-bz** [comment]
Yeah I won't be touching the CommandBuilder so we should be safe.

I've got a quick question regarding `Michaels\Manager\Traits\ManagesItemsTrait`. Does this work:

``` php
$manager->add('something.nested', ['id'=>'value']);
$manager->get('something.nested.id'); // returns "value"
```

It isn't an example in the documentation so I was wondering.

**Comment by electricjones** [comment]
yes, for as many levels as you like. Haven't tested the upper threshold, but its whatever a standard array holds.

**Comment by electricjones** [comment]
Do note that there are 3 components of Manager

`ManagesItemsTrait` adds all the methods `get(), set(), clear(), etc`
'ChainsNestedItemsTrait`adds the ability to use magic methods.
the`Manager` class uses both of the these and adds the ability to use the class as an array

**Comment by electricjones** [comment]
Closing this with fixes on #11
